module/mempool/herocache/backdata/heropool/pool_test.go

```diff
@@ -374,52 +373,49 @@ func testInvalidatingTail(t *testing.T, pool *Pool, entities []*unittest.MockEnt
 			//
 			// used tail should move backward after each invalidation
 			require.Equal(t, entities[size-i-2].ID(), usedTail.id)
-			require.Equal(t, EIndex(size-i-2), pool.used.tail.getSliceIndex())
+			require.Equal(t, EIndex(size-i-2), pool.used.tail)
 
 			// used head must point to the first element in the pool,
 			require.Equal(t, entities[0].ID(), usedHead.id)
-			require.Equal(t, EIndex(0), pool.used.head.getSliceIndex())
+			require.Equal(t, EIndex(0), pool.used.head)
 		} else {
 			// pool is empty
 			// used head and tail must be nil and their corresponding
 			// pointer indices must be undefined.
 			require.Nil(t, usedHead)
 			require.Nil(t, usedTail)
-			require.True(t, pool.used.tail.isUndefined())
-			require.True(t, pool.used.head.isUndefined())
```

Comment:
Shoudn't these two conditions be like the following? 
```go
			require.Zero(t, pool.used.tail)
			require.Zero(t, pool.used.head)
```

Reply:
If the size of a list is 0 then it's head and tail are considered as invalid and may hold any value. I think this would be the best approach. Otherwise, we would have to implement a logic to have them set to 0s and then check if this logic works. However as 0 is a completely valid index as any other it cannot bring value later on. I mean we couldn't check later for 0 and drew any conclusion on the validity out of that. 
I put "// If state has 0 size, its tail's and head's prev and next are treated as invalid " in linked list.go line 13. 

I would propose to add " If the size of a list is 0 then it's head and tail are considered as invalid and may hold any value." and not check for zeros specifically.

Please let me know if you agree. If not I'll implement a logic to set those values to 0's and will add tests for that.

Reply:
Please refer to this comment: https://github.com/onflow/flow-go/pull/4470#discussion_r1235717689

Reply:
Implemented, as check for invalid index, closing the thread.